text/3634-scoped-impl-trait-for-type.md

```diff
@@ -0,0 +1,3436 @@
+| Model overloading, mixed into nested scopes | Strict shadowing | Strict shadowing is easier to reason about for developers (especially when writing macros!), as the search stops at the nearest matching implementation or module boundary.<br>See Rust's trait method resolution behaviour and [interaction-with-specialisation] for how this is still practically compatible with a form of overload resolution.<br>See [scoped-fallback-implementations] for a possible future way to better enable adaptive behaviour in macro output. |
+| - | (Rust) Trait objects | 𝒢 does not appear to support runtime polymorphism beyond function pointers. Scoped `impl Trait for Type` is seamlessly compatible with `dyn Trait` coercions (iff `Trait` is object-safe). |
+| (unclear?) | Available implementations on discretised type parameters become part of the type identity of implementation-aware generics. | This allows code elsewhere to access scoped implementations that are already available at the definition site, and leads to overall more semantically consistent behaviour. |
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## "global" implementations
+
+I'm not too sure about the "global" wording. *Technically* that implementation isn't available for method calls unless the trait is in scope... though it is available when resolving generics. Maybe "unscoped" is better?
+
+## Precise resolution location of *implementation environments* in function calls
+
+In macros, which function-call token should provide the resolution context from where to look for scoped `impl Trait for Type`s (in all possible cases)?
+
+This doesn't matter for `Span::call_site()` vs. `Span::mixed_site()` since scoped implementations would resolve transparently through both, but it does matter for `Span::def_site()` which should exclude them.
+
+It very much does matter if one of the opt-in mitigations for [first-party-implementation-assumptions-in-macros] is implemented.
+
+## Which `struct`s should be implementation-invariant?
+[which-structs-should-be-implementation-invariant]: #which-structs-should-be-implementation-invariant
+
+This is a tough question because, runtime behaviour difference of [of-type-erased-collections] aside, the following makes shifting a type from [implementation-aware-generics] to [implementation-invariant-generics] a compilation-breaking change:
+
+```rust
+struct Type;
+struct Generic<T>(T);
+trait Trait {}
+
+mod a {
+    use super::{Type, Generic, Trait};
+    pub use impl Trait for Type {}
+    pub type Alias = Generic<T>;
+}
+
+mod b {
+    use super::{Type, Generic, Trait};
+    pub use impl Trait for Type {}
+    pub type Alias = Generic<T>;
+}
+
+use impl Trait for a::Alias {}
+use impl Trait for b::Alias {}
+```
```

Comment:
Do we want to be able to have distinct `impl Trait for Generic<Type as X in one>` and `impl Trait for Generic<Type as X in two>`? This seems unexpected.

Reply:
It's required for backwards-compatibility, or at least having a distinct `impl Trait for Generic<Type>` that isn't available for `Generic<Type as X in one>` or `Generic<Type as X in two>` is. Existing code can make safety-critical assumptions about the exact behaviour of the former.

(Forbidding implementations for concrete types that contain captured scoped implementations outright is technically also possible, but I don't think that would helpful.)

Fitting these distinct implementation targets into Rustdoc could be interesting, though, just in terms of information arrangement.